src/selectors/TraversalSelector.h

```diff
@@ -0,0 +1,153 @@
+/**
+ * TraversalSelector.h
+ *
+ *  Created on: 6/11/18
+ *     Aauthor: F. Gratl
+ */
+
+#pragma once
+
+#include <containers/cellPairTraversals/C08Traversal.h>
+#include <containers/cellPairTraversals/CellPairTraversal.h>
+#include <containers/cellPairTraversals/SlicedTraversal.h>
+#include <array>
+#include <vector>
+#include "utils/ExceptionHandler.h"
+#include "utils/Logger.h"
+
+namespace autopas {
+
+/**
+ * Possible choices for the cell pair traversal.
+ */
+enum TraversalOptions {
+  c08 = 0,
+  sliced = 1,
+};
+
+/**
+ * Provides a way to iterate over the possible choices of TraversalOption.
+ */
+static std::vector<TraversalOptions> allTraversalOptions = {TraversalOptions::c08, TraversalOptions::sliced};
+
+/**
+ * Selector for a container traversal.
+ * @tparam ParticleCell
+ */
+template <class ParticleCell>
+class TraversalSelector {
+ public:
+  /**
+   * Constructor of the TraversalSelector class.
+   * @param dims Array with the dimension lengths of the domain.
+   * @param allowedTraversalOptions Vector of traversals the selector can choose from.
+   */
+  TraversalSelector(const std::array<unsigned long, 3> &dims,
+                    const std::vector<TraversalOptions> &allowedTraversalOptions)
+      : _dims(dims), _allowedTraversalOptions(allowedTraversalOptions) {}
+
+  /**
+   * Gets the optimal traversal for a given cell functor. If no traversal is selected yet a optimum search is started.
+   * @tparam CellFunctor
+   * @param cellFunctor
+   * @return Smartpointer to the optimal traversal.
+   */
+  template <class CellFunctor>
+  std::unique_ptr<CellPairTraversal<ParticleCell, CellFunctor>> getOptimalTraversal(CellFunctor &cellFunctor);
+
+  /**
+   * Evaluates to optimal traversal based on a given cell functor.
+   * @tparam CellFunctor
+   * @param cellFunctor
+   */
+  template <class CellFunctor>
+  void tune(CellFunctor &cellFunctor);
+
+ private:
+  template <class CellFunctor>
+  std::vector<std::unique_ptr<CellPairTraversal<ParticleCell, CellFunctor>>> generateTraversals(
+      CellFunctor &cellFunctor);
+  template <class CellFunctor>
+  std::unique_ptr<CellPairTraversal<ParticleCell, CellFunctor>> chooseOptimalTraversal(
+      std::vector<std::unique_ptr<CellPairTraversal<ParticleCell, CellFunctor>>> &traversals);
+
+  // for each encountered cell processor save the optimal traversal. The cell processor is saved through its hash
```

Comment:
I do like that idea and way of saving the optimal traversals. 
Can you actually implement a test, on whether a saved optimal traversal is actually reused?

Reply:
I do not really know how since the returned traversals are always new objects. Therefore I would need to check if indeed the else branch in `getOptimalTraversal` is selected because it would obviously return the same traversal for the same type of cellFunctor even when going through the whole selection process again. Can this be tested via mocks?
At least when looking into it with gdb it works as intended :)